core/controllers/base_test.py

```diff
@@ -697,3 +697,57 @@ def test_controller_class_names(self):
                                         msg=error_message)
 
         self.assertGreater(num_handlers_checked, 150)
+
+
+class SignUpTests(test_utils.GenericTestBase):
+
+    def test_error_is_raised_on_opening_new_tab_during_signup(self):
+        """ Test that error is raised if user opens a new tab
+        during signup.
+        """
+        self.login('abc@example.com')
+        response = self.get_html_response(feconf.SIGNUP_URL)
+        csrf_token = self.get_csrf_token_from_response(response)
+
+        os.environ['USER_ID'] = ''
```

Comment:
Hi @bansalnitish, I still want to find a way to do this that does not set os.environ, since that is supposed to be private to the test framework. However, I don't understand your previous response:

> > I'm not sure you need to set os.environ directly. Wouldn't just fetching a different page work (e.g. make a GET request to /about)?
>
> I tried making a get request using get_html_response but that gets redirected since the user is not logged in and the variables are not set to initial variables due to the redirect.

When I follow the instructions in the original issue, I believe I am:

- logging in 
- navigating to the signup page
- in a new window, navigating to the About/Splash page
- completing the POST request on the signup page

What is different between the above sequence, and redirecting in the test? If get_html_response() does not reset the initial variables in the third step (but this happens if the sequence above is done manually), do you know why that is the case? In general user interactions ought to be reproducible by the test framework, so the discrepancy might point to a bigger problem here.

Reply:
Hi @seanlip,

I tried finding the reason for this. As I mentioned in the previous comment, there is a redirect when a request is made for some page i.e.

On making a request as `self.get_html_response('/about')`, I get an error `AppError: Bad response: 302 Moved Temporarily (not 200)`.

I am not sure why this redirect is happening. I tried checking by starting a local server and using requests to obtain the about page as `requests.get('http://localhost:8181/about`). The status code is 200 for the fetched response i.e. no redirect happens but in the test the status code is 302. 

May you please help me to proceed here, Thanks!

Reply:
302 seems reasonable. Find the URL that is redirected to (I think it's a property somewhere in the response) and then retrieve that URL. That should help figure out what's going on.

Reply:
The redirect is to this location: `https://www.google.com/accounts/Logout?continue=http%3A//localhost/about`

Reply:
OK. So, now: do this in the test rather than setting os.environ:

- GET /about and verify that its status code is 302 and that the redirect is what you expect
- Follow the redirect (GET the URL it redirects to). I think there is a follow() method in webtest that allows you to do that.
- Continue the rest of the test.

Reply:
I tried following the redirect but something strange happens, there are multiple redirects i.e:

```
redirect_response = self.get_html_response('about', expected_status_int=302)
full_response = redirect_response.follow()
```

This is again a redirect to: `https://www.google.com/accounts/Logout?continue=https%3A//www.google.com/accounts/Logout%3Fcontinue%3Dhttp%253A//localhost/about`.

If a do a further follow i.e `full_response.follow()`, it is again a redirect to `https://www.google.com/accounts/Logout?continue=https%3A//www.google.com/accounts/Logout%3Fcontinue%3Dhttps%253A//www.google.com/accounts/Logout%253Fcontinue%253Dhttp%25253A//localhost/about
`.

Reply:
Ah thanks -- looking at existing tests in the codebase that use follow(), it looks like this might well be the general behaviour. I suggest that you do a follow(), verify that you end up at a logout url, and then actually just go ahead and do self.logout(). Don't manipulate os.environ.

Once you've done this and fixed everything else up, let me know and I'll do a full review pass. Thanks!